### PR TITLE
ivy+docs-infra: compile angular.io with Ivy

### PR DESCRIPTION
**Pull request by gkalpak**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

## PR Type
- [x] Bugfix
- [x] Build related changes
- [x] CI related changes
- [x] angular.io application / infrastructure changes

## What is the new behavior?
Being able to build the angular.io app with Ivy and having a CircleCI job that does that.
(This is a prerequisite for running the app and tests with Ivy.)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Previously blocking issues:

- [x] _Fixed by #27697._
    Lazy routes are currently ignored by `ngtsc`: [program.ts#L133][1]
    _This is not breaking compilation, but results in only a subset of the app being compiled._
- [x] _Fixed by #27766._
    The compiler generates non-unique template function names (according to the nesting path of the element): [template.ts#L787-L788][2]
    This affects the [search-results][3] component, which happens to have two instances of `div > ul > li > a > span *ngIf` in its template:
    ```
    ERROR in ./src/app/shared/search-results/search-results.component.ts 60:9
    Module parse failed: Identifier '_div_li_span_Template_2' has already been declared (60:9)
    You may need an appropriate loader to handle this file type.
    |     }
    | }
    > function _div_li_span_Template_2(rf, ctx) {
    |     if (rf & 1) {
    |         i0.ɵelementStart(0, "span");
    ```
- [x] _Fixed by #27051._
    We may emit TS code into JS files. 

- [x] _Fixed by #28054._
    Includes a workaround for [FW-786][4], although this is now also resolved.

[1]: https://github.com/angular/angular/blob/15e671ec5a7527dae9c006073ff6b25e85142e16/packages/compiler-cli/src/ngtsc/program.ts#L133
[2]: https://github.com/angular/angular/blob/15e671ec5a7527dae9c006073ff6b25e85142e16/packages/compiler/src/render3/view/template.ts#L787-L788
[3]: https://github.com/angular/angular/blob/15e671ec5a7527dae9c006073ff6b25e85142e16/aio/src/app/shared/search-results
[4]: https://angular-team.atlassian.net/browse/FW-786

### CODE REVIEW

File: packages/compiler/src/render3/view/template.ts [hunk 1]
```diff
@@ -784,8 +784,8 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
 
     const contextName = elName ? `${this.contextName}_${elName}` : '';
 
-    const templateName =
-        contextName ? `${contextName}_Template_${templateIndex}` : `Template_${templateIndex}`;
+    const templateName = this.constantPool.uniqueName(
+        contextName ? `${contextName}_Template_${templateIndex}_` : `Template_${templateIndex}_`);
```
Comment (gkalpak) [inline packages/compiler/src/render3/view/template.ts]: Do we need a similar change for [listener functions][1]?

[1]: https://github.com/angular/angular/blob/1df4367ce83397bbcbdbc1814b07d912ae13c826/packages/compiler/src/render3/view/template.ts#L1039

**Comment by aniruddhadas9** [comment]
I am also getting the same error with ivy. Although I tried to fix it by adding `<span>` on top of it or assigning an `id` attribute to the html filed, but still having the issue.


I have created a branch to reproduce this issue and particular line of code cause the problem is as in this link

https://github.com/aniruddhadas9/candifood-ui/blob/feature/ivy-template-parsing-issue/projects/candifood/core/src/lib/components/header/header.component.html#L45:L51

The errors is as below:

```
ERROR in ./projects/candifood/core/src/lib/components/header/header.component.ts 112:9
Module parse failed: Identifier 'HeaderComponent_ul_li_Template_1' has already been declared (112:9)
You may need an appropriate loader to handle this file type.
| var _c28 = ["id", "rollUpProblemFix"];
| var _c29 = ["nav", "navbar-nav", "ml-auto", "w-100", "justify-content-end", 1, "nav", true, "navbar-nav", true, "ml-auto", true, "w-100", true, "justify-content-end", true];
> function HeaderComponent_ul_li_Template_1(rf, ctx) {
|     if (rf & 1) {
|         i0.ɵelementStart(0, "li", _c0);
ERROR in ./src/app/restaurant/components/photos/photos.component.ts 53:296
Module parse failed: Unexpected token (53:296)
You may need an appropriate loader to handle this file type.
|     PhotosComponent.prototype.ngOnInit = function () {
|     };
>     PhotosComponent.ngComponentDef = i0.ɵdefineComponent({ type: PhotosComponent, selectors: [["cfs-photos"]], factory: function PhotosComponent_Factory(t) { return new (t || PhotosComponent)(i0.ɵdirectiveInject(i1.NgbCarousel)); }, inputs: { photos: "photos" }, features: [i0.ɵProvidersFeature(())], consts: 1, vars: 1, template: function PhotosComponent_Template(rf, ctx) { if (rf & 1) {
|             i0.ɵtemplate(0, PhotosComponent_div_Template_0, 7, 1, null, _c0);
|         } if (rf & 2) {
```

**Comment by gkalpak** [comment]
@aniruddhadas9, this issue should be fixed by ced7f1d56.

**Comment by petebacondarwin** [comment]
* `refactor(docs-infra): get rid of `on` component input ` commit is no longer needed thanks to #28054.

**Review by petebacondarwin — APPROVED** [review]
LGTM - you should be able to remove that one workaround commit.

**Comment by gkalpak** [comment]
> `refactor(docs-infra): get rid of on component input` commit is no longer needed thanks to #28054.

I'd prefer to keep it anyway, because I think it is a good change.

**Review by petebacondarwin — APPROVED** [review]

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
http://test/OCL:230813412:BASE:230813487:1548376526150:743db552

**Review by IgorMinar — APPROVED** [review]
Awesome work!

**Comment by gkalpak** [comment]
merge-assistance: Requires g3sync (or maybe it doesn't: https://github.com/angular/angular/pull/26947#issuecomment-457411499)

File: aio/scripts/build-with-ivy.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+set('-e');
+
+// Constants
+const ROOT_DIR = resolve(__dirname, '..');
+const TS_CONFIG_PATH = join(ROOT_DIR, 'tsconfig.json');
+const NG_COMPILER_OPTS = {
+  angularCompilerOptions: {
+    // Related Jira issue: FW-737
+    allowEmptyCodegenFiles: true,
```
Comment (IgorMinar) [inline aio/scripts/build-with-ivy.js]: I missed this before. This is an ok workaround for now, but we should work on removing this flag.

Comment (IgorMinar) [reply aio/scripts/build-with-ivy.js]: see https://github.com/angular/angular-cli/pull/13524#pullrequestreview-196812022 for more context.

**Pull request closed without merging**